mcs/class/System/System.Net/WebOperation.cs

```diff
@@ -0,0 +1,342 @@
+		TaskCompletionSource<WebRequestStream> requestTask;
+		TaskCompletionSource<WebRequestStream> requestWrittenTask;
+		TaskCompletionSource<WebResponseStream> responseTask;
+		TaskCompletionSource<bool> completeResponseReadTask;
+		TaskCompletionSource<(bool, WebOperation)> finishedTask;
+		WebRequestStream writeStream;
+		WebResponseStream responseStream;
+		ExceptionDispatchInfo disposedInfo;
+		ExceptionDispatchInfo closedInfo;
+		WebOperation priorityRequest;
+		volatile bool finishedReading;
+		int requestSent;
+
+		public bool Aborted {
+			get {
+				if (disposedInfo != null || Request.Aborted)
+					return true;
+				if (cts != null && cts.IsCancellationRequested)
+					return true;
+				return false;
+			}
+		}
+
+		public bool Closed {
+			get {
+				return Aborted || closedInfo != null;
+			}
+		}
+
+		public void Abort ()
+		{
+			var (exception, disposed) = SetDisposed (ref disposedInfo);
+			if (!disposed)
+				return;
+			cts?.Cancel ();
+			SetCanceled ();
+			Close ();
+		}
+
+		public void Close ()
+		{
+			var (exception, closed) = SetDisposed (ref closedInfo);
+			if (!closed)
+				return;
+
+			var stream = Interlocked.Exchange (ref writeStream, null);
+			if (stream != null) {
+				try {
+					stream.Close ();
+				} catch { }
+			}
+		}
+
+		void SetCanceled ()
+		{
+			requestTask.TrySetCanceled ();
+			requestWrittenTask.TrySetCanceled ();
+			responseTask.TrySetCanceled ();
+			completeResponseReadTask.TrySetCanceled ();
+		}
+
+		void SetError (Exception error)
+		{
+			requestTask.TrySetException (error);
+			requestWrittenTask.TrySetException (error);
+			responseTask.TrySetException (error);
+			completeResponseReadTask.TrySetException (error);
+		}
+
+		(ExceptionDispatchInfo, bool) SetDisposed (ref ExceptionDispatchInfo field)
+		{
+			var wexc = new WebException (SR.GetString (SR.net_webstatus_RequestCanceled), WebExceptionStatus.RequestCanceled);
+			var exception = ExceptionDispatchInfo.Capture (wexc);
+			var old = Interlocked.CompareExchange (ref field, exception, null);
+			return (old ?? exception, old == null);
+		}
+
+		internal void ThrowIfDisposed ()
+		{
+			ThrowIfDisposed (CancellationToken.None);
+		}
+
+		internal void ThrowIfDisposed (CancellationToken cancellationToken)
+		{
+			if (Aborted || cancellationToken.IsCancellationRequested)
+				ThrowDisposed (ref disposedInfo);
+		}
+
+		internal void ThrowIfClosedOrDisposed ()
+		{
+			ThrowIfClosedOrDisposed (CancellationToken.None);
+		}
+
+		internal void ThrowIfClosedOrDisposed (CancellationToken cancellationToken)
+		{
+			if (Closed || cancellationToken.IsCancellationRequested)
+				ThrowDisposed (ref closedInfo);
+		}
+
+		void ThrowDisposed (ref ExceptionDispatchInfo field)
+		{
+			var (exception, disposed) = SetDisposed (ref field);
+			if (disposed)
+				cts?.Cancel ();
+			exception.Throw ();
+		}
+
+		internal void RegisterRequest (ServicePoint servicePoint, WebConnection connection)
+		{
+			lock (this) {
```

Comment:
Do you really need to lock arguments checks?

Reply:
Moved argument checks out.